lighthouse-core/gather/computed/page-dependency-graph.js

```diff
@@ -34,52 +38,202 @@ class PageDependencyGraphArtifact extends ComputedArtifact {
+      const evt = traceOfTab.mainThreadEvents[i];
+
+      // Skip all trace events that aren't schedulable tasks with sizable duration
+      if (
+        evt.name !== TracingProcessor.SCHEDULABLE_TASK_TITLE ||
+        !evt.dur ||
+        evt.dur < MINIMUM_TASK_DURATION
+      ) {
+        i++;
+        continue;
+      }
+
+      // Capture all events that occurred within the task
+      const children = [];
+      i++; // Start examining events after this one
+      for (
+        const endTime = evt.ts + evt.dur;
+        i < traceOfTab.mainThreadEvents.length && traceOfTab.mainThreadEvents[i].ts < endTime;
+        i++
+      ) {
+        children.push(traceOfTab.mainThreadEvents[i]);
+      }
+
+      nodes.push(new CPUNode(evt, children));
+    }
+
+    return nodes;
+  }
+
+  /**
+   * @param {!Node} rootNode
+   * @param {!NetworkNodeOutput} networkNodeOutput
+   */
+  static linkNetworkNodes(rootNode, networkNodeOutput) {
+    networkNodeOutput.nodes.forEach(node => {
+      const initiators = PageDependencyGraphArtifact.getNetworkInitiators(node.record);
       if (initiators.length) {
         initiators.forEach(initiator => {
-          const parent = urlToNodeMap.get(initiator) || rootNode;
+          const parentCandidates = networkNodeOutput.urlToNodeMap.get(initiator) || [rootNode];
+          // Only add the initiator relationship if the initiator is unambiguous
+          const parent = parentCandidates.length === 1 ? parentCandidates[0] : rootNode;
           parent.addDependent(node);
         });
-      } else if (record !== rootRequest) {
+      } else if (node !== rootNode) {
         rootNode.addDependent(node);
       }
     });
+  }
+
+  /**
+   * @param {!Node} rootNode
+   * @param {!NetworkNodeOutput} networkNodeOutput
+   * @param {!Array<!CPUNode>} cpuNodes
+   * @param {!TraceOfTabArtifact} traceOfTab
+   */
+  static linkCPUNodes(rootNode, networkNodeOutput, cpuNodes) {
+    function addDependentNetworkRequest(cpuNode, reqId) {
+      const networkNode = networkNodeOutput.idToNodeMap.get(reqId);
+      if (!networkNode || networkNode.resourceType !== 'xhr') return;
+      networkNode.addDependency(cpuNode);
+    }
+
+    function addDependencyOnUrl(cpuNode, url) {
+      const candidates = networkNodeOutput.urlToNodeMap.get(url) || [];
+
+      let minCandidate = null;
+      let minDistance = Infinity;
+      // Find the closest request that finished before this CPU task started
+      candidates.forEach(candidate => {
+        const distance = cpuNode.startTime - candidate.endTime;
+        if (distance > 0 && distance < minDistance) {
+          minCandidate = candidate;
+          minDistance = distance;
+        }
+      });
+
+      if (!minCandidate) return;
+      cpuNode.addDependency(minCandidate);
+    }
+
+    const timers = new Map();
+    for (const node of cpuNodes) {
+      for (const evt of node.children) {
+        if (!evt.args.data) continue;
+
+        const url = evt.args.data.url;
```

Comment:
seems like it's possible for this to be undefined in some cases? (does `ScheduleStyleRecalculation` always a url? or EvaluateScript for eval()?)
i guess you could early return in `addDependencyOnUrl` to avoid some work?

Reply:
yep will do